eth/downloader/queue.go

```diff
@@ -54,15 +60,59 @@ type fetchRequest struct {
 // fetchResult is a struct collecting partial results from data fetchers until
 // all outstanding pieces complete and the result as a whole can be processed.
 type fetchResult struct {
-	Pending int         // Number of data fetches still pending
-	Hash    common.Hash // Hash of the header to prevent recalculating
+	pending int32 // Flag telling what deliveries are outstanding
 
 	Header       *types.Header
 	Uncles       []*types.Header
 	Transactions types.Transactions
 	Receipts     types.Receipts
 }
 
+func newFetchResult(header *types.Header, fastSync bool) *fetchResult {
+	item := &fetchResult{
+		Header: header,
+	}
+	if !header.EmptyBody() {
+		item.pending = 1
+	}
+	if fastSync && !header.EmptyReceipts() {
+		item.pending += 2
+	}
+	return item
+}
+
+// SetBodyDone flags the body as finished.
+func (f *fetchResult) SetBodyDone() {
+	if v := atomic.LoadInt32(&f.pending); v == 1 || v == 3 {
+		atomic.AddInt32(&f.pending, -1)
+	}
+}
+
+// AllDone checks item is done
+func (f *fetchResult) AllDone() bool {
+	return atomic.LoadInt32(&f.pending) == 0
+}
+
+// SetReceiptsDone flags the receipts as finished.
+func (f *fetchResult) SetReceiptsDone() {
+	if v := atomic.LoadInt32(&f.pending); v == 2 || v == 3 {
+		atomic.AddInt32(&f.pending, -2)
+	}
+}
+
+// CheckDone checks if the given type is done already
+func (f *fetchResult) Done(typ int) bool {
+	v := atomic.LoadInt32(&f.pending)
+	switch typ {
+	case BodyType:
+		return !(v == 1 || v == 3)
```

Comment:
This check would be more obvious if you test the bit instead of comparing against 1 or 3.

`v & 0x1 == 1`